react/Gallery.js

```diff
@@ -59,62 +62,81 @@ const Gallery = ({
 
     return (
       <CellMeasurer
-        cache={cache}
+        cache={this.cache}
         columnIndex={0}
         key={key}
         parent={parent}
         rowIndex={index}
       >
-        {({ measure }) => (
-          <div className={containerClasses} key={key} style={style} onLoad={measure}>
-            {rowItems.map(renderItem)}
-          </div>
-        )}
+        <div className={containerClasses} key={key} style={style}>
+          {rowItems.map(this.renderItem)}
+        </div>
       </CellMeasurer>
     )
   }
 
-  const ssrContainer = classNames(`${searchResult.gallery} pa3 bn`, {
-    [searchResult.galleryTwoColumns]: layoutMode == 'small' && mobile
-  })
+  handleSSR() {
+    const { products, layoutMode, runtime: { hints: { mobile } } } = this.props
 
-  const ssrFallBack = (
-    <div className={ssrContainer}>
-      {products.map(renderItem)}
-    </div>
-  )
-
-  return (
-    <NoSSR onSSR={ssrFallBack}>
-      <div className={`${searchResult.gallery} pa3 bn`}>
-        <Adopt
-          mapper={{
-            scroller: <WindowScroller />,
-            autoSizer: <AutoSizer disableHeight />,
-          }}
-          mapProps={({ scroller: { height }, autoSizer: { width } }) => ({ width, height })}
-        >
-          {({ width, height }) => {
-            const itemsPerRow = (layoutMode === 'small' && mobile) ? TWO_ITEMS : (Math.floor(width / itemWidth) || ONE_ITEM)
-            const nRows = Math.ceil(products.length / itemsPerRow)
-
-            return (
-              <List
-                key={layoutMode}
-                deferredMeasurementCache={cache}
-                autoHeight
-                width={width}
-                height={height}
-                rowCount={nRows}
-                rowHeight={cache.rowHeight}
-                rowRenderer={args => renderRow({ ...args, itemsPerRow })}
-              />
-            )
-          }}
-        </Adopt>
-      </div >
-    </NoSSR>
-  )
+    const ssrContainer = classNames(`${searchResult.gallery} pa3 bn`, {
+      [searchResult.galleryTwoColumns]: layoutMode === 'small' && mobile,
+    })
+
+    return (
+      <div className={ssrContainer}>
+        {products.map(this.renderItem)}
+      </div>
+    )
+  }
+
+  render() {
+    const {
+      products,
+      layoutMode,
+      runtime: { hints: { mobile } },
+      itemWidth,
+    } = this.props
+
+    // Reset the cache to recalculate the heights when layout changes
+    if (layoutMode !== this.state.prevLayoutMode) {
+      this.cache.clearAll()
+    }
+
+    return (
+      <NoSSR onSSR={this.handleSSR()}>
+        <div className={`${searchResult.gallery} pa3 bn`}>
+          <Adopt
+            mapper={{
+              scroller: <WindowScroller />,
+              autoSizer: <AutoSizer disableHeight />,
+            }}
+            mapProps={({ scroller: { height, scrollTop, isScrolling }, autoSizer: { width } }) => ({ width, height, scrollTop, isScrolling })}
```

Comment:
You could put `mapProps` value on a constant outside return block for better readability, what do you think?

Reply:
Yes, thanks for the suggestion.